File: app/src/main/java/com/rendersoncs/reportform/login/RemoveUserActivity.java
package com.rendersoncs.reportform.login;

import android.os.Bundle;
import android.view.View;
import android.widget.EditText;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.crashlytics.android.Crashlytics;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.EmailAuthProvider;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.ValueEventListener;
import com.rendersoncs.reportform.R;
import com.rendersoncs.reportform.login.util.User;

import io.fabric.sdk.android.Fabric;

public class RemoveUserActivity extends AppCompatActivity
        implements ValueEventListener, DatabaseReference.CompletionListener {

    private User user;
    private EditText password;
    private FirebaseAuth mAuth;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_remove_user);
        Fabric.with(this, new Crashlytics());

        mAuth = FirebaseAuth.getInstance();
    }

    @Override
    protected void onResume() {
        super.onResume();
        init();
    }

    private void init(){
        password = findViewById(R.id.password);

        user = new User();
        user.setId( mAuth.getCurrentUser().getUid() );
        user.contextDataDB( this );
    }

    public void removeUser( View view ){
        if (password.getText().toString().isEmpty()) {
            Toast.makeText(this, "Por favor insira a senha.", Toast.LENGTH_SHORT).show();
        } else {
            user.setPassword(password.getText().toString());

            reauthenticate();
        }
    }

    private void reauthenticate(){
        FirebaseUser firebaseUser = mAuth.getCurrentUser();

        if( firebaseUser == null ){
            return;
        }

        AuthCredential credential = EmailAuthProvider.getCredential(
                user.getEmail(),
                user.getPassword()
        );

        firebaseUser.reauthenticate( credential )
                .addOnCompleteListener(task -> {

                    if( task.isSuccessful() ){
                        deleteUser();
                    }
                }).addOnFailureListener(e -> {
                    Crashlytics.logException( e );
                    Toast.makeText(
                            RemoveUserActivity.this,
                            e.getMessage(),
                            Toast.LENGTH_SHORT
                    ).show();
                });
    }

    private void deleteUser(){
        FirebaseUser firebaseUser = mAuth.getCurrentUser();

        if( firebaseUser == null ){
            return;
        }

        firebaseUser.delete().addOnCompleteListener(task -> {

            if( !task.isSuccessful() ){
                return;
            }

            user.removeDB( RemoveUserActivity.this );
        })
                .addOnFailureListener(e -> {
                    Crashlytics.logException( e );
                    Toast.makeText(
                            RemoveUserActivity.this,
                            e.getMessage(),
                            Toast.LENGTH_SHORT
                    ).show();
                });
    }

    @Override
    public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
        User u = dataSnapshot.getValue( User.class );
        if (u != null) {
            user.setEmail( u.getEmail() );
        }
    }

    @Override
    public void onCancelled(@NonNull DatabaseError firebaseError) {
        Crashlytics.logException( firebaseError.toException() );
    }

    @Override
    public void onComplete(DatabaseError databaseError,@NonNull DatabaseReference databaseReference) {
        if( databaseError != null ){
            Crashlytics.logException( databaseError.toException() );
        }

        Toast.makeText(
                RemoveUserActivity.this,
                getResources().getString(R.string.label_account_removed),
                Toast.LENGTH_SHORT
        ).show();
        finish();
    }
}

File: app/src/main/java/com/rendersoncs/reportform/view/activitys/login/RemoveUserActivity.kt
package com.rendersoncs.reportform.view.activitys.login

import android.os.Bundle
import android.view.KeyEvent
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.crashlytics.android.Crashlytics
import com.google.android.gms.tasks.Task
import com.google.firebase.auth.EmailAuthProvider
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.ValueEventListener
import com.rendersoncs.reportform.R
import com.rendersoncs.reportform.view.activitys.login.util.User
import com.rendersoncs.reportform.view.services.util.closeVirtualKeyBoard
import io.fabric.sdk.android.Fabric
import kotlinx.android.synthetic.main.activity_remove_user.*

class RemoveUserActivity : AppCompatActivity(), ValueEventListener,
        DatabaseReference.CompletionListener,
        TextView.OnEditorActionListener {

    private var user: User? = null
    //private var password: EditText? = null
    private var mAuth: FirebaseAuth? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_remove_user)
        Fabric.with(this, Crashlytics())
        mAuth = FirebaseAuth.getInstance()
        password.setOnEditorActionListener(this)
    }

    override fun onResume() {
        super.onResume()
        init()
    }

    private fun init() {
        //password = findViewById(R.id.password)
        user = User()
        user!!.id = mAuth!!.currentUser!!.uid
        user!!.contextDataDB(this)
    }

    fun btnRemoveUser(view: View?) {
        this.removeUser()
    }

    private fun removeUser() {
        if (password!!.text.toString().isEmpty()) {
            Toast.makeText(this, resources.getString(R.string.label_insert_password), Toast.LENGTH_SHORT).show()
        } else {
            user!!.password = password!!.text.toString()
            reAuthenticate()
        }
    }

    private fun reAuthenticate() {
        val firebaseUser = mAuth!!.currentUser ?: return
        val credential = EmailAuthProvider.getCredential(
                user!!.email,
                user!!.password
        )
        firebaseUser.reauthenticate(credential)
                .addOnCompleteListener { task: Task<Void?> ->
                    if (task.isSuccessful) {
                        deleteUser()
                    }
                }.addOnFailureListener { e: Exception ->
                    Crashlytics.logException(e)
                    Toast.makeText(
                            this@RemoveUserActivity,
                            e.message,
                            Toast.LENGTH_SHORT
                    ).show()
                }
    }

    private fun deleteUser() {
        val firebaseUser = mAuth!!.currentUser ?: return
        firebaseUser.delete().addOnCompleteListener { task: Task<Void?> ->
                    if (!task.isSuccessful) {
                        return@addOnCompleteListener
                    }
                    user!!.removeDB(this@RemoveUserActivity)
                }
                .addOnFailureListener { e: Exception ->
                    Crashlytics.logException(e)
                    Toast.makeText(
                            this@RemoveUserActivity,
                            e.message,
                            Toast.LENGTH_SHORT
                    ).show()
                }
    }

    override fun onDataChange(dataSnapshot: DataSnapshot) {
        val u = dataSnapshot.getValue(User::class.java)
        if (u != null) {
            user!!.email = u.email
        }
    }

    override fun onCancelled(fireBaseError: DatabaseError) {
        Crashlytics.logException(fireBaseError.toException())
    }

    override fun onComplete(databaseError: DatabaseError?, databaseReference: DatabaseReference) {
        if (databaseError != null) {
            Crashlytics.logException(databaseError.toException())
        }
        Toast.makeText(
                this@RemoveUserActivity,
                resources.getString(R.string.label_account_removed),
                Toast.LENGTH_SHORT
        ).show()
        finish()
    }

    override fun onEditorAction(view: TextView, actionId: Int, event: KeyEvent?): Boolean {
        if (actionId == EditorInfo.IME_ACTION_DONE) {
            closeVirtualKeyBoard(
                    this,
                    view
            )
            this.removeUser()
            return true
        }
        return false
    }
}